bundledatasrc.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/juju/errors"
+	"gopkg.in/yaml.v2"
+)
+
+type fieldPresenseMap map[interface{}]interface{}
+
+func (fpm fieldPresenseMap) fieldPresent(fieldName string) bool {
+	_, exists := fpm[fieldName]
+	return exists
+}
+
+func (fpm fieldPresenseMap) forField(fieldName string) fieldPresenseMap {
+	v, exists := fpm[fieldName]
+	if !exists {
+		return nil
+	}
+
+	asMap, valid := v.(fieldPresenseMap)
+	if !valid {
+		panic(errors.Errorf("field map entry %q does not point to a nested field presense map", fieldName))
+	}
+	return asMap
+}
+
+// bundleDataPart combines a parsed BundleData instance with a nested map that
+// can be used to discriminate between fields that are missing from the data
+// and those that are present but defined to be empty.
+type bundleDataPart struct {
+	data        *BundleData
+	presenseMap fieldPresenseMap
+}
+
+// BundleDataSource is implemented by types that can parse bundle data into a
+// list of composable parts.
+type BundleDataSource interface {
+	read() ([]*bundleDataPart, error)
+}
+
+// FileBundleDataSource is a BundleDataSource that reads bundle data from a file.
+type FileBundleDataSource string
+
+// read implements BundleDataSource.
+func (s FileBundleDataSource) read() ([]*bundleDataPart, error) {
+	f, err := os.Open(string(s))
+	if err != nil {
+		return nil, errors.Annotatef(err, "open file bundle data source")
+	}
+	defer func() { _ = f.Close() }()
+
+	return NewStreamBundleDataSource(f).read()
+}
+
+// StreamBundleDataSource is a BundleDataSource that reads bundle data from an
+// io.Reader instance.
+type StreamBundleDataSource struct {
+	r io.Reader
+}
+
+// NewStreamBundleDataSource returns a StreamBundleDataSource instance that
+// reads data from r.
+func NewStreamBundleDataSource(r io.Reader) StreamBundleDataSource {
+	return StreamBundleDataSource{r: r}
+}
+
+// read implements BundleDataSource.
+func (s StreamBundleDataSource) read() ([]*bundleDataPart, error) {
+	b, err := ioutil.ReadAll(s.r)
+	if err != nil {
+		return nil, errors.Annotatef(err, "file bundle data source (%v)", s)
+	}
+
+	var (
+		// Ideally, we would be using a single reader and we would
+		// rewind it to read each block in structured and raw mode.
+		// Unfortunately, the yaml parser seems to parse all documents
+		// at once so we need to use two decoders.
+		structDec = yaml.NewDecoder(bytes.NewReader(b))
+		rawDec    = yaml.NewDecoder(bytes.NewReader(b))
+		parts     []*bundleDataPart
+	)
+
+	for docIdx := 0; ; docIdx++ {
+		var part bundleDataPart
+
+		err = structDec.Decode(&part.data)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, errors.Annotatef(err, "unmarshal document %d from %v", docIdx, s)
+		}
+
+		// We have already checked for errors for the previous unmarshal attempt
+		_ = rawDec.Decode(&part.presenseMap)
```

Comment:
This feels a lot like it should actually be a change to the YAML Decode contract that would allow us to decode into a struct, but also ask the decoder what attributes were actually there. We can live with this, but what do you think about trying to update Decode to something like DecodeWithKeys

Reply:
Interesting idea. I also thought it would be great not to have to do the decode dance twice. I believe we can land the current set of changes, then open a PR for the yaml package and address this shortcoming in the future. What do you think?

Reply:
works for me